### PR TITLE
New: Add no-whitespace-before-property rule (fixes #1086)

### LINKED ISSUE #1086: Rule: no-space-before-property
If expression is inline, whitespace should not be allowed after dot or before square brackets.

valid:

```
foo[bar]

foo
  .bar()
  .baz()

foo.
  bar().
  baz()
```

invalid:

```
foo [bar]

foo. bar. baz
```

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/3104722-rule-no-space-before-property?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by kaicataldo**

### CODE REVIEW

**Comment by Dashed** [comment]
woah. awesome. looking forward to this. 

thanks @kaicataldo!  :+1:

File: docs/rules/no-whitespace-before-property.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+  .baz
+
+foo
+  .bar. baz
+```
+
+The following patterns are not considered problems:
+
+```js
+/*eslint no-whitespace-before-property: 2*
```
Comment (gyandeeps) [inline docs/rules/no-whitespace-before-property.md]: missing a `/` at the end.

File: lib/rules/no-whitespace-before-property.js [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+                    context.report({
+                        node: node,
+                        message: "Unexpected whitespace before property \"{{ propName }}\".",
+                        data: {
+                            propName: prop.name
+                        }
+                    });
+                }
+            }
+            return;
```
Comment (gyandeeps) [inline lib/rules/no-whitespace-before-property.js]: Return is implicit here, you dont have to say this.

**Comment by kaicataldo** [comment]
Thanks for reviewing, @gyandeeps!

**Comment by kaicataldo** [comment]
Realized I forgot to include tests with tabs - just updated my PR.

**Comment by gyandeeps** [comment]
Your build s failing because message is missing `#` before the issue number. Expected `(fixes #XXXX)`

**Comment by kaicataldo** [comment]
Yeah, thanks - had to run out he door right after making my PR but will fix when I'm back at my computer.

File: docs/rules/no-whitespace-before-property.md [hunk 3]
```diff
@@ -0,0 +1,53 @@
+# Disallow whitespace before properties (no-whitespace-before-property)
+
+Disallows whitespace around dot or brackets before properties of objects if they are on the same line.
```
Comment (nzakas) [inline docs/rules/no-whitespace-before-property.md]: Please add more here. This section should explain the comfort that this rule is treating, not the rule itself. See http://eslint.org/docs/rules/comma-dangle as an example.

File: docs/rules/no-whitespace-before-property.md [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+## Rule Details
+
```
Comment (nzakas) [inline docs/rules/no-whitespace-before-property.md]: Need a sentence here in the form "This rule aims to (what does it do) and as such will report a problem when (when does it report a problem)."

**Comment by kaicataldo** [comment]
Updated the docs - thanks for the direction, makes sense. Let me know if anything else needs to be improved.

**Comment by mysticatea** [comment]
Could this rule catch the following pattern?
If no, should this rule catch these?

```js
people
    .  hello();

people  .
    hello();
```

**Comment by kaicataldo** [comment]
@mysticatea It currently doesn't catch that pattern, as it ignores multiline member expressions. It would catch this:

```
people . hello()    // the whitespace on this line would error
  . helloAgain();   // but the whitespace on this line would not

**Comment by platinumazure** [comment]
@mysticatea @kaicataldo If we decide to take those on, make sure to watch for conflicts with `dot-location`.

**Comment by nzakas** [comment]
Let's not add extra conditions now. We can always add later if necessary.

File: docs/rules/no-whitespace-before-property.md [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+JavaScript allows whitespace between objects and their properties. However, inconsistent spacing can make code harder to read and can lead to errors.
+
+```js
+foo. bar .baz . quz
+```
+
+## Rule Details
+
+This rule alerts for whitespace around the dot or brackets before properties of objects if they are on the same line. It does not alert for whitespace when the object and property are on separate lines, as it is common to add newlines to longer chains of properties to improve readability:
```
Comment (nzakas) [inline docs/rules/no-whitespace-before-property.md]: "Readability" is personal and not everyone agrees on what is readable. I'd just drop that part from the sentence.

File: lib/rules/no-whitespace-before-property.js [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    //--------------------------------------------------------------------------
+    // Public
+    //--------------------------------------------------------------------------
+
+    return {
+        MemberExpression: function(node) {
+            var obj = node.object;
+            var prop = node.property;
+
+            if (astUtils.isTokenOnSameLine(obj, prop)) {
```
Comment (nzakas) [inline lib/rules/no-whitespace-before-property.js]: Better to use the `SourceCode` version of this method, which checks for comments as well.

Comment (nzakas) [inline lib/rules/no-whitespace-before-property.js]: Oops, nevermind. Confused this method with something else.

File: docs/rules/no-whitespace-before-property.md (hunk 5 above)
Comment (kaicataldo) [inline docs/rules/no-whitespace-before-property.md]: Sounds good

File: lib/rules/no-whitespace-before-property.js (hunk 6 above)
Comment (kaicataldo) [inline lib/rules/no-whitespace-before-property.js]: Updated!

**Comment by kaicataldo** [comment]
@nzakas Sounds good re: not adding extra conditions now. I'm happy to look at this rule again in the future should we decide we want this. :+1:

**Comment by kaicataldo** [comment]
If we decide we want to add more conditions in the future, I don't think it would be difficult to make the rule do a more granular check by checking for whitespace between the punctuators and their respective objects/properties, rather than the whitespace between objects and properties.

**Comment by gyandeeps** [comment]
LGTM. Thanks @kaicataldo

**Pull request merged**